src/types/basic.rs

```diff
@@ -270,22 +272,147 @@ pub struct TransactionIndex {
 
 pub type AggregateSigPairing = id::constants::IpPairing;
 
+/// FIXME: Move higher up in the dependency
+#[derive(SerdeBase16Serialize, Serialize, Debug)]
+pub struct BakerAggregationSignKey {
+    pub sign_key: aggregate_sig::SecretKey<AggregateSigPairing>,
+}
+
+impl BakerAggregationSignKey {
+    pub fn generate<T: CryptoRng + RngCore>(csprng: &mut T) -> Self {
+        Self {
+            sign_key: aggregate_sig::SecretKey::generate(csprng),
+        }
+    }
+
+    pub fn prove<T: CryptoRng + RngCore>(
+        &self,
+        csprng: &mut T,
+        random_oracle: &mut RandomOracle,
+    ) -> aggregate_sig::Proof<AggregateSigPairing> {
+        self.sign_key.prove(csprng, random_oracle)
+    }
+}
+
 /// FIXME: Move higher up in the dependency
 #[derive(SerdeBase16Serialize, Serialize, Clone, Debug)]
 pub struct BakerAggregationVerifyKey {
     pub verify_key: aggregate_sig::PublicKey<AggregateSigPairing>,
 }
 
+impl BakerAggregationVerifyKey {
+    pub fn from_sign_key(secret: &BakerAggregationSignKey) -> Self {
+        Self {
+            verify_key: aggregate_sig::PublicKey::from_secret(&secret.sign_key),
+        }
+    }
+}
+
+/// FIXME: Move higher up in the dependency
+#[derive(SerdeBase16Serialize, Serialize, Debug)]
+pub struct BakerSignatureSignKey {
+    pub sign_key: ed25519_dalek::SecretKey,
+}
+
+impl BakerSignatureSignKey {
+    pub fn generate<T: CryptoRng + RngCore>(csprng: &mut T) -> Self {
+        Self {
+            sign_key: ed25519_dalek::SecretKey::generate(csprng),
+        }
+    }
+}
+
 /// FIXME: Move higher up in the dependency
 #[derive(SerdeBase16Serialize, Serialize, Clone, Debug)]
-pub struct BakerSignVerifyKey {
+pub struct BakerSignatureVerifyKey {
     pub verify_key: ed25519_dalek::PublicKey,
 }
 
+impl BakerSignatureVerifyKey {
+    pub fn from_sign_key(secret: &BakerSignatureSignKey) -> Self {
+        Self {
+            verify_key: ed25519_dalek::PublicKey::from(&secret.sign_key),
+        }
+    }
+}
+
+/// FIXME: Move higher up in the dependency
+#[derive(SerdeBase16Serialize, Serialize, Debug)]
+pub struct BakerElectionSignKey {
+    pub sign_key: ecvrf::SecretKey,
```

Comment:
Does this really need to be `pub`?

I'd also recommend maybe not having `Debug` since this is secret data.

Reply:
making the proofs for the BakerKeysPayload needs this, so I have set it to pub(crate) instead